sycl/doc/design/OffloadDesign.md

```diff
@@ -0,0 +1,326 @@
+
+To reduce the potential size of the linked device binary, an additional host
+link step is performed to gather dependency information when static device
+libraries are being compiled.  This information is sent through the
+`clang-offload-deps` tool to generate a dependency IR file which is used
+during the device link step.
+
+The use of the clang-linker-wrapper introduces the support of LTO for
+device code.  We can leverage this and move away from the dependency gathering
+information step with 'clang-offload-deps' and use thinLTO for device code.
+
+There are multiple device linking steps that also occur.  The first step links
+together all of the objects and the required device libraries.  The second is
+performed including all of optional device libraries, the static device
+libraries and the dependency information that was gathered above.  This link
+step is performed with `--only-needed` to streamline the final device binary.
+
+The device libraries that are pulled in during the device link step will need
+to be controlled by the `clang-linker-wrapper`.  There is a controlling option
+(`-fno-sycl-device-lib=arg`) that is available to the user which needs to be
+provided to the wrapper to give more control over what device libraries are
+pulled in during the device link.
+
+### Ahead Of Time Compilation
+
+The updated offloading model will integrate the Ahead of Time (AOT) compilation
+behaviors into the clang-linker-wrapper.  The actions will typically take place
+after the device link and post link steps.
+
+Regardless of the AOT target, the flow is similar, only modifying the offline
+compiler that is used to create the target device image.  It is expected that
+the offline compiler will also use unique command lines specific to the tool
+to create the image.
+
+To support the needed option passing triggered by use of the
+`-Xsycl-target-backend` option and implied options based on the optional
+device behaviors for AOT compilations for GPU new command line interfaces
+are needed to pass along this information.
+
+| Target | Triple        | Offline Tool   | Option for Additional Args |
+|--------|---------------|----------------|----------------------------|
+| CPU    | spir64_x86_64 | opencl-aot     | `--cpu-tool-arg=<arg>`     |
+| GPU    | spir64_gen    | ocloc          | `--gen-tool-arg=<arg>`     |
+| FPGA   | spir64_fpga   | aoc/opencl-aot | `--fpga-tool-arg=<arg>`    |
+
+*Table: Ahead of Time Info*
+
+To complete the support needed for the various targets using the
+`clang-linker-wrapper` as the main interface, a few additional options will
+be needed to communicate from the driver to the tool.  Further details of usage
+are given further below.
+
+| Option Name                  | Purpose                                      |
+|------------------------------|----------------------------------------------|
+| `--fpga-tool-deps=<arg>`     | Comma separated list of dependency files used for FPGA hardware compiles using `aoc` |
```

Comment:
Is this list only ever needed for FPGA compile? Or can it be generalized to something like `--backend-tool-deps` and populated only when necessary?

Reply:
As I think of this, the dependency file that is generated during the compilation should be added to the final binary before the link is called.  This would mean that this option is not necessary.  It also implies we need to call out the ability to add the dependency file to the fat binary so it can be used during the link/call to aoc.